src/feature/hs/hs_common.c

```diff
@@ -1723,45 +1714,38 @@ hs_get_extend_info_from_lspecs(const smartlist_t *lspecs,
     }
   } SMARTLIST_FOREACH_END(ls);
 
+  /* Choose a preferred address first, but fall back to an allowed address.
+   * Skip IPv6 if we're not direct as relays only extend via IPv4. */
+  fascist_firewall_choose_address_ls(lspecs, 0, &ap, direct_conn);
+
   /* Legacy ID is mandatory, and we require IPv4. */
-  if (!have_v4 || !have_legacy_id) {
+  if (!tor_addr_port_is_valid_ap(&ap, 0) || !have_legacy_id) {
     goto done;
   }
 
-  /* We know we have IPv4, because we just checked. */
+  /* We know we have an IP address, because we just checked. */
   if (!direct_conn) {
     /* All clients can extend to any IPv4 via a 3-hop path. */
     goto validate;
-  } else if (direct_conn &&
```

Comment:
Don't delete this "else if" block, because its conditions are always true at this point.
Instead, turn it into an "else" block.